### DISCUSSION TITLE
Express js middleware’s integration error with apollov4

### DISCUSSION THREAD

**Discussion by SwadipSinghoRoy** [apollographql/apollo-server#7778] [Q&A] [answered]
I’m using Express js with @apollo/server & installed various middleware e.g: morgan, cors, cookieParser. But they didn’t executed.

```
// required dependencies here
..........
// Multiple schema configured file
const mergedSchema = require('./schemas/schemaConfig');

// Apollo server connection build
const apolloserver = new ApolloServer({
  schema: mergedSchema
})

// connection establishing
async function serverStart(server) {
  const PORT = process.env.PORT || 2080
  try {
    const { url } = await startStandaloneServer(server, {
      context: ({req, res}) => ({req, res}),
      listen: { port: PORT }
    })
    console.log(`# ~Server ready at: ${url}`);
    
    await mongoose.connect('mongodb://127.0.0.1:27017/DB-Name', {
      useNewUrlParser: true
    })
    mongoose.set("strictQuery", false);
    console.log("~Database connected.....");
  } catch (error) {
    console.log(error);
  }
}
serverStart(apolloserver);

// applied middlewares
app.use(
  morgan('dev'),
  cors(),
  express.json(),
  cookieParser()
)
```
Applied middleware doesn’t executing e.g: Didn’t executed morgan logger, cookieParser() etc. I tried `expressMiddleware()` with `await server.start()` instead of `await startStandaloneServer()` for managing middleware but facing another issues. When i’m going to `await startStandaloneServer()` approach server started & working, but **middleware’s not working expectedly**.

**Comment by glasser** [comment]
I'm a bit confused by your code snippet above. It doesn't match the text below.

startStandaloneServer is for when you don't want to tweak middleware with Express.  If you want to do that, use `expressMiddleware`. It seems like you understand that, but the code snippet doesn't show what you tried to do with `expressMiddleware`.

Ideally, a full reproduction (try codesandbox.io) is helpful.

**Reply by SwadipSinghoRoy** [reply]
I was trying to manage used `startStandaloneServer`. As well as I tried `expressMiddleware` But facing another error. 
```
Error: You must `await server.start()` before calling `expressMiddleware()`
    at ApolloServer.assertStarted
```
My applied context of code is:
```
const express = require('express');
const { ApolloServer } = require('@apollo/server');
const mongoose = require('mongoose');
const morgan = require('morgan');
const cors = require('cors');
const cookieParser = require('cookie-parser');
const { expressMiddleware } = require('@apollo/server/express4');

const mergedSchema = require('./schemas/schemaConfig');

const PORT = process.env.PORT || 5080;
const app = express();

// Apollo server connection
const server = new ApolloServer({
  schema: mergedSchema
})

async function serverStart() {
  try {
    await server.start()
    mongoose.connect('mongodb://127.0.0.1:27017/graphql-express-apollo', {
      useNewUrlParser: true
    })
    mongoose.set("strictQuery", false);
    console.log("~Database connected.....");
  } catch (error) {
    console.log(error);
  }
}

serverStart()

app.use(
  '/graphql',
  morgan('dev'),
  cors(),
  express.json(),
  cookieParser(),
  expressMiddleware(server, {
    context: ({req, res}) => ({req, res})
  })
)

app.listen(PORT, () => {
  console.log(`Server listening port - ${PORT}`);
})
```

> According to requirements, `await server.start()` is executed first and this is executed by `serverStart()`. Then `expressMiddleware()` will be executed.

But getting error. I don't understand how apply these middleware's.This is not possible to work with extarnal middleware when use `startStandaloneServer`?

**Reply by glasser** [reply]
You're calling the async function serverStart without awaiting it, which means that the next line (app.use) is not actually after server.start() finishes.

**Reply by SwadipSinghoRoy** [reply]
I was `awaiting serverStart()` but getting same error. Updated code snippet:
```
// Required dependency,
// Required multiple schema config file,
// Apollo connection

async function serverStart() {
  try {
    await server.start()
    console.log("Initialized server");
  } catch (error) {
    console.log(error);
  }
}

serverStart().then(() => {
  mongoose.connect('mongodb://127.0.0.1:27017/DB-Name', {
      useNewUrlParser: true
  })
  mongoose.set("strictQuery", false);
  console.log("~Database connected.....");
}).catch((err) => {
  console.log(err);
})

app.use(
  '/graphql',
  morgan('dev'),
  cors(),
  express.json(),
  cookieParser(),
  expressMiddleware(server, {
    context: ({req, res}) => ({req, res})
  })
) 

app.listen(PORT, () => {
  console.log(`Server listening port - ${PORT}`);
})
```
I noticed that `app` executed before executing `serverStart()`. So what will be the corrected step?

**Reply by glasser** [reply]
If you can't use top level await, you can put the whole rest of your program (app.use etc) inside the function you pass to `then`.

**Reply by SwadipSinghoRoy** [reply]
That means **DB connection, middleware** passing everything put on `then` block. If yes, I tried this & server was running but `morgan logger` behave unexpectedly. Suppose i'm sending query request server responding, But continue to print the same request log until a new request is sent.

**Reply by trevor-scheer** [reply]
Everything @glasser has suggested so far is correct. It's hard for us to answer why your logger would be behaving unexpectedly without having a reproduction we can use. If you can provide a codesandbox that demonstrates this behavior we can be helpful.

**Reply by SwadipSinghoRoy** [reply]
I applied as @glasser said. [Sandbox implementation](https://codesandbox.io/p/sandbox/intelligent-haibt-cqqwzg?file=%2Findex.js%3A1%2C36&layout=%257B%2522sidebarPanel%2522%253A%2522EXPLORER%2522%252C%2522rootPanelGroup%2522%253A%257B%2522direction%2522%253A%2522horizontal%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522id%2522%253A%2522ROOT_LAYOUT%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522cloq20nd20007356l19gf2b95%2522%252C%2522sizes%2522%253A%255B70%252C30%255D%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522EDITOR%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522id%2522%253A%2522cloq20nd20003356lwqksgbh6%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522SHELLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522id%2522%253A%2522cloq20nd20005356lm1qhzkxq%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522DEVTOOLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522id%2522%253A%2522cloq20nd20006356l3mf9f0iw%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%252C%2522sizes%2522%253A%255B63.96514161220044%252C36.03485838779956%255D%257D%252C%2522tabbedPanels%2522%253A%257B%2522cloq20nd20003356lwqksgbh6%2522%253A%257B%2522id%2522%253A%2522cloq20nd20003356lwqksgbh6%2522%252C%2522tabs%2522%253A%255B%257B%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252Findex.js%2522%252C%2522id%2522%253A%2522cloq22dmd00mg356l760sc7cm%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%255D%252C%2522activeTabId%2522%253A%2522cloq22dmd00mg356l760sc7cm%2522%257D%252C%2522cloq20nd20006356l3mf9f0iw%2522%253A%257B%2522id%2522%253A%2522cloq20nd20006356l3mf9f0iw%2522%252C%2522activeTabId%2522%253A%2522cloq2kt9i01mh356l0ls1q0z3%2522%252C%2522tabs%2522%253A%255B%257B%2522type%2522%253A%2522UNASSIGNED_PORT%2522%252C%2522port%2522%253A5080%252C%2522id%2522%253A%2522cloq2kt9i01mh356l0ls1q0z3%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522path%2522%253A%2522%252F%2522%257D%255D%257D%252C%2522cloq20nd20005356lm1qhzkxq%2522%253A%257B%2522id%2522%253A%2522cloq20nd20005356lm1qhzkxq%2522%252C%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522cloq20nd20004356l466pgiom%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522TASK_LOG%2522%252C%2522taskId%2522%253A%2522start%2522%257D%255D%252C%2522activeTabId%2522%253A%2522cloq20nd20004356l466pgiom%2522%257D%257D%252C%2522showDevtools%2522%253Atrue%252C%2522showShells%2522%253Atrue%252C%2522showSidebar%2522%253Atrue%252C%2522sidebarPanelSize%2522%253A15%257D)

**Reply by trevor-scheer** [reply]
The sandbox was broken, I fixed it (`resolver` -> `resolvers`).
It seems to be working, but I don't see any logs at all and I'm not sure why (not even console logs)

[sandbox](https://codesandbox.io/p/sandbox/magical-worker-d6lkqm?file=%2F.codesandbox%2Ftasks.json%3A1%2C1&layout=%257B%2522sidebarPanel%2522%253A%2522EXPLORER%2522%252C%2522rootPanelGroup%2522%253A%257B%2522direction%2522%253A%2522horizontal%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522id%2522%253A%2522ROOT_LAYOUT%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522cloq20nd20007356l19gf2b95%2522%252C%2522sizes%2522%253A%255B70%252C30%255D%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522EDITOR%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522id%2522%253A%2522cloq20nd20003356lwqksgbh6%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522SHELLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522id%2522%253A%2522cloq3t4q000o83b6h73ec6z6w%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522DEVTOOLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522id%2522%253A%2522cloq20nd20006356l3mf9f0iw%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%252C%2522sizes%2522%253A%255B42.32221019577113%252C57.67778980422887%255D%257D%252C%2522tabbedPanels%2522%253A%257B%2522cloq20nd20003356lwqksgbh6%2522%253A%257B%2522id%2522%253A%2522cloq20nd20003356lwqksgbh6%2522%252C%2522activeTabId%2522%253A%2522cloq22dmd00mg356l760sc7cm%2522%252C%2522tabs%2522%253A%255B%257B%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252Findex.js%2522%252C%2522id%2522%253A%2522cloq22dmd00mg356l760sc7cm%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%252C%257B%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252Fpackage.json%2522%252C%2522id%2522%253A%2522cloq3oviu00453b6ha2lzsyim%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%252C%257B%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252F.codesandbox%252Ftasks.json%2522%252C%2522id%2522%253A%2522cloq3ud4b013n3b6hm9urp41d%2522%252C%2522mode%2522%253A%2522temporary%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%255D%257D%252C%2522cloq20nd20006356l3mf9f0iw%2522%253A%257B%2522id%2522%253A%2522cloq20nd20006356l3mf9f0iw%2522%252C%2522activeTabId%2522%253A%2522cloq2kt9i01mh356l0ls1q0z3%2522%252C%2522tabs%2522%253A%255B%257B%2522type%2522%253A%2522UNASSIGNED_PORT%2522%252C%2522port%2522%253A5080%252C%2522id%2522%253A%2522cloq2kt9i01mh356l0ls1q0z3%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522path%2522%253A%2522%252F%2522%257D%255D%257D%252C%2522cloq3t4q000o83b6h73ec6z6w%2522%253A%257B%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522cloq3sz9p00o43b6haxd0m993%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522TERMINAL%2522%252C%2522shellId%2522%253A%2522cloq3t4x6018mefg66d0tboqp%2522%257D%255D%252C%2522id%2522%253A%2522cloq3t4q000o83b6h73ec6z6w%2522%252C%2522activeTabId%2522%253A%2522cloq3sz9p00o43b6haxd0m993%2522%257D%257D%252C%2522showDevtools%2522%253Atrue%252C%2522showShells%2522%253Atrue%252C%2522showSidebar%2522%253Atrue%252C%2522sidebarPanelSize%2522%253A15%257D)

**Reply by SwadipSinghoRoy** [reply]
Everything is working but don't print log so i can't understand what is happening actually. But log printed in my local device.

**Reply by trevor-scheer** [reply, marked as answer]
Ok. I ran your code locally and logging was fine. Maybe something with codesandbox.

The reason you're seeing the repeated logs seemingly after each query is because Explorer is querying your schema every second for changes. If you close Explorer or disable schema polling you'll notice the logs stop.

**Reply by SwadipSinghoRoy** [reply]
I don't understand, how explorer is querying automatically without send request from client side & how occured changes? Any suggested documentation/guideline for closing or disabling schema polling?

**Reply by trevor-scheer** [reply]
Explorer runs an _introspection_ query which asks the graph about the schema itself. Explorer uses information about the schema to provide useful features like autocomplete.

To disable in the UI:
https://www.apollographql.com/docs/graphos/explorer/additional-features#local-development
To disable via config:
https://www.apollographql.com/docs/apollo-server/api/plugin/landing-pages#pollforschemaupdates

**Reply by trevor-scheer** [reply]
It _does_ send a request, which you can see in the network tab in your browser.
